_posts/2019-06-18-questions-every-unit-test-should-answer.md

```diff
@@ -0,0 +1,198 @@
+---
+layout: post
+title: Questions every unit test should answer
+date: 2019-06-25 12:13:14 +0200
+categories: testing
+author_name: Jacek Smolak
+author_url : /author/jaceksmolak
+author_avatar: jaceksmolak
+read_time : 5
+show_related_posts: false
+square_related:
+excerpt: "We should write tests to prevent defects from happening. That is one of the key roles tests play. But not everyone knows there is more to writing tests than that."
+feature_image: posts/2019-06-18-questions-every-unit-test-should-answer/poster.jpg
+---
+
+We should write tests to prevent defects from happening. That is one of the key roles tests play. But not everyone knows there is more to writing tests than that. Tests also explain two very important questions behind the feature they, well, test. Those are: the _how_ and the _why_.
+
+## The 'how'
+
+But not how the code is written, rather than how it is working. I find it the single most important question your tests must answer. 
+
+ - How is this piece of code working?
+ - What to expect given certain input?
+ - What is happening if something goes wrong?
+ - How it will behave under certain conditions?
+
+The how, if answered properly, will tell you all about it.
+
+Let's take [Fizz Buzz](https://en.wikipedia.org/wiki/Fizz_buzz) as an example. A simple game, with even simpler rules. But is it? Not so while ago I conducted, with a colleague of mine, an interview for an engineer position for our company. Fizz Buzz came out, and I asked: given that you know how it works, how many tests would you write to explain the rules (yes, that was a trick question). The answer was: around 50. I asked why? Why not 100? OK—perhaps 60, 70 (was the answer). The person interviewed clearly did not understand what I wanted to achieve.
+
+What is the number of tests, the maximum number of tests, Fizz Buzz requires in order to answer the how behind it?
+
+…
+
+It's 4. All you need is four tests, to get to know how Fizz Buz works. Here's the list:
+
+```javascript
+it('should return Fizz when the number is divisible by 3', () => {
```

Comment:
I would change the wording to contain "only"
`should return Fizz when the number is divisible only by 3`
As you know in this game given number can be divisible by 3 and 5 then the output is different.
I like that you named the variable `anyNumberDivisibleBy3` which made the concrete number irrelevant. But in my view this means that we could use 15 as an example, that should fail the test though, because it's divisible by 5 too.

Also what do you think about adding more assertions in one test case?

Reply:
Hmm... if we go this way, 3 is divisible by 3 and by 1 (by 2 as well, but with a change). It's hard to come up with a short and yet complete name. Perhaps `anyNumberDivisibleBy3AndNotBy5` or similar?

More assertions... well, I had a talk with Wolfram about it, but I will break it down here.
Let's take the new name, `anyNumberDivisibleBy3AndNotBy5`. How would I name such variable if I were to use it more than one? OK - let's say, that has been solved.
Now I have two or more:
```javascript
const number1 = 3;
const number2 = 6;
const number3 = 9;
```
Why 3 numbers, why not 4, or 5 or 42? How many is enough?
If you will think about those tests as a TDD way of implementing the code, having one will not produce the very same code, as intended. Rather than something like:
```javascript
fuzzBuzz() { return 'fizz'; }
```
as that `return 'fizz';` would pass a test with even twenty assertions.

But let's say, having more assertions would bring us to TDD'ing the implementation, and it would be as we'd intend it (doing the division, without change, returning a number, or fizz / buzz / fizzbuzz). You'd be left with multiple assertions using multiple numbers. Wouldn't that be a bit overkill to understand _how_ the unit under the test works? I guess you'd want to refactor the tests (that are code as well, and need taking care of and being easy to read and understand), and most likely you'd do something with those 20, 30 or more assertions.

That is why I used `any` in the name. Any number that is divisible by 3 will have that function return `fizz`. Any number divisible by 5 returns `buzz`. And finally, any number divisible by both 3 and 5 will return `fizzbuzz`. And passed number for all other cases.

`any` covers here all of the cases, by being well described. And if you'd want to test more numbers - you'd still have the possibility, by using  property based testing, for which this is a very good case.

Reply:
I like the name `anyNumberDivisibleOnlyBy3`, I think it would also require changing the test description to __should return Fizz when the number is divisible only by 3__. Just to be consistent. What do you think?

As for multiple assertions. I understand your point. I quess my question was motivated by having naive understanding of the reason why TDD is used for. Implementing only enough to pass the tests.